File: lab_00/main.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>

#define OK 0
#define ERROR 1
#define ZERO_DEF -1

void clean_buf()
{
    char a = getchar();
    while (a != '\n' && a != EOF)
        a = getchar();
}

void free_mat(long double ***arr, int n) {
    for (int i = 0; i < n; i++)
        free((*arr)[i]);
    free(*arr);
}

long double **create_mat(int n, int *err) {
    *err = OK;
    long double **arr = calloc(n, sizeof(long double *));
    if (arr == NULL) {
        *err = ERROR;
    } else {
        for (int i = 0; i < n && *err == OK; i++) {
            arr[i] = calloc(n, sizeof(long double));
            if (arr[i] == NULL) {
                *err = ERROR;
                free_mat(&arr, i);
            }
        }
    }
    if (*err == ERROR)
        printf("Проблемы с выделением памяти, Босс(\n");
    return arr;
}

int read_int(char str[]) {
    printf("%s", str);
    int flag = 1;
    int n;
    while (flag) {
        if (scanf("%d", &n) != 1 || n < 0)
        {
            printf("Ошибка ввода! Повторите попытку ввода целого неотрицательного числа: ");
        } else {
            flag = 0;
        }
        clean_buf();
    }
    return n + 1;
}

long double read_double(char str[]) {
    printf("%s", str);
    long double n;
    while (scanf("%Lf", &n) != 1)
    {
        clean_buf();
        printf("Ошибка ввода! Повторите попытку ввода вещественного числа: ");
    }
    clean_buf();
    //printf("%Lf\n", n);
    return n;
}

void print_matrix(long double **mat, int size) {
    printf("Матрица:\n");
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++)
            printf("%.15Lf ", mat[i][j]);
        printf("\n");
    }
    if (size == 0)
        printf("Матрица пуста.");
    printf("\n");
}

void print_array(long double arr[], int size) {
    printf("Массив:\n");
    for (int i = 0; i < size; i++)
        printf("%.15Lf\n", arr[i]);
    if (size == 0)
        printf("Массив пуст.");
    printf("\n");
}

void print_res(long double arr[], int n) {
    if (n - 1 < 0)
        printf("Результат:\n P0(x) =");
    else
        printf("Результат:\n P%d(x) = ", n - 1);
    for (int i = 0; i < n; i++) {
        char *plus = "+ ";
        if (i == 0 || arr[n - i - 1] < 0)
            plus = "";
        if (i < (n - 1))
            printf("%s%.2Lf * x^%d ", plus, arr[n - i - 1], n - i - 1);
        else
            printf("%s%.2Lf ", plus, arr[n - i - 1]);
    }
    if (n == 0)
        printf("0");
    printf("\n");
}

int input_all(long double ***arr, long double **yrr, int *err) {
    *err = OK;
    int n = read_int("Введите степень уравнения: ");
    int count_diff = read_int("Введите известное количество производных: ");
    int count_points = ceil(((double) n) / ((double) count_diff));
    *arr = create_mat(n, err);
    if (*err == OK) {
        *yrr = calloc(n, sizeof(long double));
        if (*yrr == NULL) {
            free_mat(arr, n);
            *err = ERROR;
        } else {
            for (int j = 0; j < count_points; j++) {
                char message[50];
                long double xi;
                int *diff_arr = calloc(n, sizeof(int));
                if (diff_arr == NULL) {
                    free_mat(arr, n);
                    free(*yrr);
                    *err = ERROR;
                } else {
                    for (int k = 0; k < n; k++)
                        diff_arr[k] = 1;
                    sprintf(message, "Введите x%d : ", j + 1);
                    xi = read_double(message);
                    for (int l = 0; l < count_diff; l++)
                        for (int m = l; m < n; m++) {
                            if ((l + j * count_diff) < n)
                                (*arr)[l + j * count_diff][m] = powl(xi, (m - l)) * diff_arr[m];
                            diff_arr[m] *= m - l;
                        }
                    sprintf(message, "Введите y%d : ", j + 1);
                    (*yrr)[j * count_diff] = read_double(message);
                    //printf("TYT!\n\n");
                    for (int l = 1; l < count_diff; l++)
                        if ((l + j * count_diff) < n) {
                            sprintf(message, "Введите производную №%d : ", l);
                            (*yrr)[l + j * count_diff] = read_double(message);
                        }
                }
            }
        }
    }
    return n;
}

long double **make_minor(long double **arr, int n, int i, int *err) {
    *err = OK;
    long double **minor = create_mat(n - 1, err);
    if (*err == OK) {
        for (int row = 1; row < n; row++)
            for (int col = 0; col < n; col++)
                if (col != i)
                    minor[row - 1][col - (col > i)] = arr[row][col];
    }
    return minor;
}

long double calc_def(long double **arr, int n, int *err) {
    *err = OK;
    if (n <= 0)
        return 0;
    if (n == 1)
        return arr[0][0];
    if (n == 2)
        return arr[0][0] * arr[1][1] - arr[1][0] * arr[0][1];
    long double res = 0;
    for (int i = 0; i < n && *err == OK; i++) {
        long double **minor = make_minor(arr, n, i, err);
        if (*err == OK) {
            res += arr[0][i] * calc_def(minor, n - 1, err) * powl(-1, i);
            if (*err == OK)
                free_mat(&minor, n - 1);
        }
    }
    return res;
}

long double calc_ai(long double **arr, long double *yrr, int n, int ind, long double def_main, int *err) {
    *err = OK;
    long double def_ai = 0;
    long double **arr_new = create_mat(n, err);
    if (*err == OK) {
        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
                arr_new[i][j] = arr[i][j];
        for (int i = 0; i < n; i++)
            arr_new[i][ind] = yrr[i];
        def_ai = calc_def(arr_new, n, err);
        if (*err == OK)
            free_mat(&arr_new, n);
    }
    return def_ai / def_main;
}

int main(void) {
    int rc = OK;
    long double **arr;
    long double *yrr;
    int n = input_all(&arr, &yrr, &rc);
    if (rc == OK) {
        long double def_main = calc_def(arr, n, &rc);
        if (rc == OK) {
            if (def_main == 0) {
                printf("Система уравнений не имеет единственного решения, так как определитель матрицы равен 0.\n\
        Возможно, система имеет бесконечно много решений или не имеет решений вовсе.");
                rc = ZERO_DEF;
            } else {
                long double *res_a = calloc(n, sizeof(long double));
                if (res_a == NULL)
                    rc = ERROR;
                else {
                    for (int i = 0; i < n && rc == OK; i++) {
                        long double ai = calc_ai(arr, yrr, n, i, def_main, &rc);
                        if (rc == OK)
                            res_a[i] = ai;
                    }
                    if (rc == OK)
                        print_res(res_a, n);
                    free(res_a);
                }
            }
        }
        free_mat(&arr, n);
        free(yrr);
    }
    return rc;
}

File: lab_00/main.py
# Талышева ИУ7-45Б лаба_0 выч_алг
#from sympy import diff
import math

# читает с проверкой целое неотрицательное число с клавиатуры (возвращает число + 1)
def read_int(str):
    print(str, end = '')
    flag = True
    while flag:
        try:
            n = int(input()) + 1
            if (n <= 0):
                print("Число должно быть неотрицательным! Повторите попытку: ")
            else:
                flag = False
        except:
            print("Ошибка ввода! Повторите попытку ввода целого неотрицательного числа: ", end = "")
    return n

# читает с проверкой вещественное число с клавиатуры
def read_float(str):
    print(str, end = '')
    flag = True
    while flag:
        try:
            n = float(input())
            flag = False
        except:
            print("Ошибка ввода! Повторите попытку ввода вещественного числа: ", end = "")
    return n

# выводит матрицу на экран
def print_matrix(mat):
    print("Матрица:")
    for i in range(len(mat)):
        for j in range(len(mat[0])):
            print(mat[i][j], end = ' ')
        print()
    if len(mat) == 0 or len(mat[0]) == 0:
        print("Матрица пуста.")
    print()

# выводит массив на экран
def print_array(arr):
    print("Массив:")
    for i in range(len(arr)):
        print(arr[i])
    if len(arr) == 0:
        print("Массив пуст.")
    print()

# выводит результат в красивом виде на экран
def print_res(arr):
    n = len(arr)
    new_arr = arr[::-1]
    res_str = ""
    if n - 1 < 0:
        print("Результат:\n P0(x) =", end = " ")
    else:
        print("Результат:\n P" + str(n - 1) + "(x) =", end = " ")
    for i in range(n):
        plus = "+ "
        if res_str == "" or new_arr[i] < 0:
            plus = ""
        x_part = " * x^" + str(n - i - 1) if i < (n - 1) else ""
        if (new_arr[i] != 0):
            res_str += plus + "{:.2f}".format(new_arr[i]) + x_part + " "
    if n == 0:
        res_str += "0"
    print(res_str)

# считывает всё что нужно для задачи
def input_all():
    # вводится степнь уравнения
    n = read_int("Введите степень уравнения: ")
    count_diff = read_int("Введите известное количество производных: ")
    count_points = math.ceil(n / count_diff)
    #print(count_points)
    arr = [[0] * n for _ in range(n)]
    yrr = [0] * n
    for i in range(count_points):
        diff_arr = [1] * n
        xi = read_float("Введите x" + str(i + 1) + " : ")
        for j in range(count_diff):
            for k in range(j, n):
                #print(i, j, k)
                if (j + i * count_diff) < n:
                    arr[j + i * count_diff][k] = xi ** (k - j) * diff_arr[k]
                #print(k, j, arr[j + i * count_diff][k])
                #print_matrix(arr)
            #print("ДО:", diff_arr)
            diff_arr = [diff_arr[w] * (w - j) for w in range(n)]
            #print("После:", diff_arr)
        #print_matrix(arr)
        #print_array(yrr)
        yrr[i * count_diff] = read_float("Введите y" + str(i + 1) + " : ")
        for j in range(1, count_diff):
            if (j + i * count_diff) < n:
                yrr[j + i * count_diff] = read_float("Введите производную №" + str(j) + " : ")
    return n, arr, yrr

# вычислит определитель
def calc_def(arr):
    n = len(arr)
    # базовые случаи
    if n <= 0:
        return 0
    if n == 1:
        return arr[0][0]
    if n == 2:
        return arr[0][0] * arr[1][1] - arr[1][0] * arr[0][1]
    # для матриц размером > 2 для каждого элемента первой строки
    res = 0
    for i in range(n):
        # вычисляем минор элемента arr[0][i]
        minor = [[arr[row][col] for col in range(n) if col != i] for row in range(1, n)]
        # вычисляем определитель минора и добавляем его с учетом знака
        res += arr[0][i] * calc_def(minor) * ((-1) ** i)
    return res

# вычисляет методом Крамера i-тый неизвестный элемент
def calc_ai(arr, yrr, ind, def_main):
    arr_new = [row[:] for row in arr]
    for i in range(len(arr_new)):
        arr_new[i][ind] = yrr[i]
    def_ai = calc_def(arr_new)
    #print_matrix(arr_new)
    #print(def_ai / def_main)
    return def_ai / def_main

n, arr, yrr = input_all()
##n = 2
##arr = [[1, 2, 4], [0, 1, 4], [0, 0, 2]]
##yrr = [13, 11, 4]
#print_matrix(arr)
#print_array(yrr)
def_main = calc_def(arr)
#print(def_main)
if (def_main == 0):
    print("Система уравнений не имеет единственного решения, так как определитель матрицы равен 0.\n\
Возможно, система имеет бесконечно много решений или не имеет решений вовсе.")
else:
    res_a = list()
    for i in range(len(arr[0])):
        ai = calc_ai(arr, yrr, i, def_main)
        res_a.append(ai)
    print_res(res_a)
